mailpoet/assets/js/src/segments/dynamic/dynamic_segments_filters/fields/woocommerce/used_coupon_code.tsx

```diff
@@ -36,6 +38,111 @@ export function UsedCouponCodeFields({
     [filterIndex],
   );
   const { updateSegmentFilter } = useDispatch(storeName);
+  // isInitialized is used for displaying loading message before loading coupons
+  const [isInitialized, setIsInitialized] = useState<boolean>(false);
+  const [coupons, setCoupons] = useState<SelectOption[]>([]);
+  // additionalLoading state is used for displaying loading message when loading more coupons
+  const [additionalLoading, setAdditionalLoading] = useState<boolean>(false);
+  const [page, setPage] = useState<number>(1);
+  // hasMore state is used to prevent loading more coupons when there are no more coupons to load
+  const [hasMore, setHasMore] = useState<boolean>(true);
+  const [searchQuery, setSearchQuery] = useState<string>('');
+  const [errors, setErrors] = useState([]);
+
+  const loadCoupons = useCallback(
+    (
+      isInitialLoading: boolean,
+      newPage: number,
+      newSearchQuery: string,
+      newHasMore: boolean,
+    ) => {
+      if (!newHasMore) {
+        return;
+      }
+
+      if (!isInitialLoading) {
+        setAdditionalLoading(true);
+      }
+
+      void MailPoet.Ajax.post({
+        api_version: MailPoet.apiVersion,
+        endpoint: 'coupons',
+        action: 'getCoupons',
+        data: {
+          page_number: newPage,
+          page_size: 1000,
+          include_coupon_ids: segment.coupon_code_ids,
+          search: newSearchQuery,
+        },
+      })
+        .then((response) => {
+          const { data } = response;
+          const loadedCoupons: SelectOption[] = data.map((coupon: Coupon) => ({
+            value: coupon.id.toString(),
+            label: coupon.text,
+          }));
+          const nextPage = newPage + 1;
+          if (loadedCoupons.length === 0) {
+            setHasMore(false);
+          } else {
+            setCoupons((prevOptions: SelectOption[]) =>
+              uniqBy(
+                (coupon) => coupon.value,
+                [...prevOptions, ...loadedCoupons],
+              ),
+            );
```

Comment:
Very minor, but when I tested this out (with a lower page size since I don't have thousands of coupons), I noticed the select jumps back to the top after loading the new coupons. Probably no easy way to prevent that, and if anyone has >1000 coupons to go through they'll probably be using the search instead.